sigstore/_internal/tuf.py

```diff
@@ -0,0 +1,159 @@
+
+    These directories are not guaranteed to already exist.
+    """
+
+    builder = appdirs.AppDirs("sigstore-python", "sigstore")
+    tuf_base = parse.quote(url, safe="")
+
+    data_dir = Path(builder.user_data_dir)
+    cache_dir = Path(builder.user_cache_dir)
+
+    return (data_dir / tuf_base), (cache_dir / tuf_base)
+
+
+class TrustUpdater:
+    """Internal trust root (certificates and keys) downloader.
+
+    TrustUpdater discovers the currently valid certificates and keys and
+    securely downloads them from the remote TUF repository at 'url'.
+
+    TrustUpdater expects to find an initial root.json in either the local
+    metadata directory for this URL, or (as special case for the sigstore.dev
+    production and staging instances) in the application resources.
+    """
+
+    def __init__(self, url: str) -> None:
+        self._repo_url = url
+        self._updater: Optional[Updater] = None
+
+        self._metadata_dir, self._targets_dir = _get_dirs(url)
+
+        # intialize metadata dir
+        tuf_root = self._metadata_dir / "root.json"
+        if not tuf_root.exists():
+            if self._repo_url == DEFAULT_TUF_URL:
+                fname = "root.json"
+            elif self._repo_url == STAGING_TUF_URL:
+                fname = "staging-root.json"
+            else:
+                raise Exception(f"TUF root not found in {tuf_root}")
+
+            self._metadata_dir.mkdir(parents=True, exist_ok=True)
+            root_json = read_embedded(fname)
+            with tuf_root.open("wb") as io:
+                io.write(root_json)
+
+        # intialize targets cache dir
+        # TODO: Pre-populate with any targets we ship with sources
+        self._targets_dir.mkdir(parents=True, exist_ok=True)
+
+        logger.debug("TUF metadata: %s", self._metadata_dir)
+        logger.debug("TUF targets cache: %s", self._targets_dir)
+
+    @classmethod
+    def production(cls) -> "TrustUpdater":
+        return cls(DEFAULT_TUF_URL)
+
+    @classmethod
+    def staging(cls) -> "TrustUpdater":
+        return cls(STAGING_TUF_URL)
+
+    def _setup(self) -> "Updater":
+        """Initialize and update the toplevel TUF metadata"""
+        updater = Updater(
+            metadata_dir=str(self._metadata_dir),
+            metadata_base_url=f"{self._repo_url}",
+            target_base_url=f"{self._repo_url}targets/",
+            target_dir=str(self._targets_dir),
+            fetcher=_fetcher,
+        )
+
+        # NOTE: we would like to avoid refresh if the toplevel metadata is valid.
+        # https://github.com/theupdateframework/python-tuf/issues/2225
+        updater.refresh()
+        return updater
+
+    def _get(self, usage: str) -> List[bytes]:
+        """Return all active targets with given usage"""
+        if not self._updater:
+            self._updater = self._setup()
+
+        data = []
+        assert self._updater._trusted_set.targets
```

Comment:
`bandit` flagged this -- is there a reason for this `assert` beyond maybe an error handling case that we should use an exception for?

Reply:
A nicer solution would be great, the issue is this:
* `_trusted_set.targets` is correctly annotated `Optional[Metadata[Targets]]`
* the Optional is there because before the metadata is refreshed, `_trusted_set.targets` is None
* but at this point when metadata has been successfully refreshed we know that the actual annotation is `Metadata[Targets]`
* unfortunately mypy cannot understand this: the assert is there to inform mypy

So if you have other clever solutions, be my guest.

Reply:
if no smarter solutions appear, I would suggest a "# nosec" to disable bandit here: this feels like an appropriate use of assert to me.

Reply:
Agreed, I think that's fine.

Reply:
I think we can use `typing.cast` for this! I'll push that up.

Reply:
Ah, looks like `cast` won't quite work. But I'm going to replace this with a `# type: ignore[union-attr]`, since that silences the underlying MyPy problem rather than turning it into a Bandit problem that we then ignore 🙂

Reply:
that'll work but I'll note it's not just mypy: it's static typing in general for that code block. So e.g. editor predictions become unavailable as well when assert is removed, this is unfortunate.

I think the ultimate fix is to avoid Optional in cases like this. I will look at the python-tuf code and see what could be done.